src/tbb/misc_ex.cpp

```diff
@@ -369,6 +370,48 @@ int FindProcessorGroupIndex ( int procIdx ) {
     return i;
 }
 
+#endif
+
+/*
+* Algorithm is designed on the assumption that the workers thread id go from 1 to Hard limit set by TBB market::global_market
+* The worker threads are added starting from processor group same as the master thread and when it's full go to the next processor group.
+* If there is an oversubscription each worker thread is allocated to the processor group in round robin order.
+*/
+int FindProcessorGroupIndex(int procIdx) {
+
+    int proc_id_tracker = 0;
+    // initialize the current processor grp index to master thread processor group.
+     int current_grp_idx = ProcessorGroupInfo::HoleIndex;
+    __TBB_ASSERT(hardware_concurrency_info == do_once_state::initialized, "FindProcessorGroupIndex is used before AvailableHwConcurrency");
+
+    if (ProcessorGroupInfo::HoleIndex >= 1) {
+            // shift the worker_id index to position it on processor group same as master thread
+            proc_id_tracker = procIdx + (theProcessorGroups[ProcessorGroupInfo::HoleIndex - 1].numProcsRunningTotal);
+    } else {
+            // master thread is in group 0 => No shifting required
+            proc_id_tracker = procIdx;
+    }
+
+    if (procIdx > theProcessorGroups[ProcessorGroupInfo::NumGroups - 1].numProcsRunningTotal) {
+        // over subscription - Round Robin.
+        current_grp_idx = (current_grp_idx + 1) % (ProcessorGroupInfo::NumGroups);
```

Comment:
Does this achieve round robin? current_group_idx is a local variable that is set to HoleIndex. I think HoleIndex does not change and so the initial value of the local current_grp_idx will not change regardless of the procIdx.  And so if there is oversubscription, the current_grp_idx will always be (HoleIndex + 1)%NumGroups. Or am I missing something...?

Reply:
Discussed offline and fixed.